File: scala/layer.scala
import neuron.Neuron
import sigmoid.SigmoidNeuron

abstract class Layer {
  def neurons : Array[Neuron]
  def feedForward(inputs : Array[Double]) : Array[Double] = neurons map { x => x.feedForward(inputs) }
  def backPropagate(error : Array[Double]) : Array[Double] = (neurons, error).zipped.map{ (n,e)  => n.backPropagate(e) }.transpose.map{ _.sum }.toArray
}

File: c++/layer.cpp
//Layer implementation
// mkrum

#include "layer.h"
#include <iostream>

Layer::Layer(int size){
    for(int i = 0; i < size; i++){
        layer.push_back(Neuron(size));
    }
    output.reserve(size);
    error.reserve(size);
}

vector<double> Layer::feedForward(vector<double> inputs){
    output.erase(output.begin(), output.end());
    for(int i = 0; i < layer.size(); i++){
        output.push_back(layer[i].feedForward(inputs));
    }
    return output;
}

vector<double> Layer::backPropagate(vector<double> inError){
    vector<double> temp;
    std::fill(error.begin(), error.end(), 0);
    for(int i = 0; i < inError.size(); i++){
        temp = layer[0].backPropagate(inError[0]);
        for(int i = 0; i < temp.size(); i++){
            error[i] += temp[i];
        }
    }
    return error;
}
